File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/PipelineUtils.java
/*
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.cloud.firestore;

import static com.google.cloud.firestore.pipeline.expressions.AggregateFunction.countAll;
import static com.google.cloud.firestore.pipeline.expressions.Expr.and;
import static com.google.cloud.firestore.pipeline.expressions.Expr.arrayContainsAny;
import static com.google.cloud.firestore.pipeline.expressions.Expr.eqAny;
import static com.google.cloud.firestore.pipeline.expressions.Expr.field;
import static com.google.cloud.firestore.pipeline.expressions.Expr.not;
import static com.google.cloud.firestore.pipeline.expressions.Expr.or;
import static com.google.cloud.firestore.pipeline.expressions.FunctionUtils.aggregateFunctionToValue;
import static com.google.cloud.firestore.pipeline.expressions.FunctionUtils.exprToValue;

import com.google.api.core.InternalApi;
import com.google.cloud.firestore.Query.ComparisonFilterInternal;
import com.google.cloud.firestore.Query.CompositeFilterInternal;
import com.google.cloud.firestore.Query.FilterInternal;
import com.google.cloud.firestore.Query.LimitType;
import com.google.cloud.firestore.Query.UnaryFilterInternal;
import com.google.cloud.firestore.pipeline.expressions.AggregateFunction;
import com.google.cloud.firestore.pipeline.expressions.AliasedAggregate;
import com.google.cloud.firestore.pipeline.expressions.AliasedExpr;
import com.google.cloud.firestore.pipeline.expressions.BooleanExpr;
import com.google.cloud.firestore.pipeline.expressions.Expr;
import com.google.cloud.firestore.pipeline.expressions.Field;
import com.google.cloud.firestore.pipeline.expressions.Selectable;
import com.google.common.collect.Lists;
import com.google.firestore.v1.Cursor;
import com.google.firestore.v1.MapValue;
import com.google.firestore.v1.Value;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@InternalApi
public class PipelineUtils {
  @InternalApi
  public static Value encodeValue(Object value) {
    return UserDataConverter.encodeValue(FieldPath.empty(), value, UserDataConverter.ARGUMENT);
  }

  @InternalApi
  public static Value encodeValue(Expr value) {
    return exprToValue(value);
  }

  @InternalApi
  public static Value encodeValue(AggregateFunction value) {
    return aggregateFunctionToValue(value);
  }

  @InternalApi
  public static Value encodeValue(String value) {
    return Value.newBuilder().setStringValue(value).build();
  }

  @InternalApi
  public static Value encodeValue(boolean value) {
    return Value.newBuilder().setBooleanValue(value).build();
  }

  @InternalApi
  public static Value encodeValue(long value) {
    return Value.newBuilder().setIntegerValue(value).build();
  }

  @InternalApi
  public static Value encodeValue(double value) {
    return Value.newBuilder().setDoubleValue(value).build();
  }

  @InternalApi
  public static Value encodeValue(Map<String, Value> options) {
    return Value.newBuilder()
        .setMapValue(MapValue.newBuilder().putAllFields(options).build())
        .build();
  }

  @InternalApi
  static BooleanExpr toPipelineBooleanExpr(FilterInternal f) {
    if (f instanceof ComparisonFilterInternal) {
      ComparisonFilterInternal comparisonFilter = (ComparisonFilterInternal) f;
      Field field = Field.ofServerPath(comparisonFilter.fieldReference.getFieldPath());
      Value value = comparisonFilter.value;
      switch (comparisonFilter.operator) {
        case LESS_THAN:
          return and(field.exists(), field.lt(value));
        case LESS_THAN_OR_EQUAL:
          return and(field.exists(), field.lte(value));
        case GREATER_THAN:
          return and(field.exists(), field.gt(value));
        case GREATER_THAN_OR_EQUAL:
          return and(field.exists(), field.gte(value));
        case EQUAL:
          return and(field.exists(), field.eq(value));
        case NOT_EQUAL:
          return and(field.exists(), not(field.eq(value)));
        case ARRAY_CONTAINS:
          return and(field.exists(), field.arrayContains(value));
        case IN:
          List<Value> valuesList = value.getArrayValue().getValuesList();
          return and(field.exists(), eqAny(field, Lists.newArrayList(valuesList)));
        case ARRAY_CONTAINS_ANY:
          List<Value> valuesListAny = value.getArrayValue().getValuesList();
          return and(field.exists(), arrayContainsAny(field, Lists.newArrayList(valuesListAny)));
        case NOT_IN:
          List<Value> notInValues = value.getArrayValue().getValuesList();
          return and(field.exists(), not(eqAny(field, Lists.newArrayList(notInValues))));
        default:
          // Handle OPERATOR_UNSPECIFIED and UNRECOGNIZED cases as needed
          throw new IllegalArgumentException("Unsupported operator: " + comparisonFilter.operator);
      }
    } else if (f instanceof CompositeFilterInternal) {
      CompositeFilterInternal compositeFilter = (CompositeFilterInternal) f;
      switch (compositeFilter.getOperator()) {
        case AND:
          List<BooleanExpr> conditions =
              compositeFilter.getFilters().stream()
                  .map(PipelineUtils::toPipelineBooleanExpr)
                  .collect(Collectors.toList());
          return and(
              conditions.get(0),
              conditions.subList(1, conditions.size()).toArray(new BooleanExpr[0]));
        case OR:
          List<BooleanExpr> orConditions =
              compositeFilter.getFilters().stream()
                  .map(PipelineUtils::toPipelineBooleanExpr)
                  .collect(Collectors.toList());
          return or(
              orConditions.get(0),
              orConditions.subList(1, orConditions.size()).toArray(new BooleanExpr[0]));
        default:
          // Handle OPERATOR_UNSPECIFIED and UNRECOGNIZED cases as needed
          throw new IllegalArgumentException(
              "Unsupported operator: " + compositeFilter.getOperator());
      }
    } else if (f instanceof UnaryFilterInternal) {
      UnaryFilterInternal unaryFilter = (UnaryFilterInternal) f;
      Field field = Field.ofServerPath(unaryFilter.fieldReference.getFieldPath());
      switch (unaryFilter.getOperator()) {
        case IS_NAN:
          return and(field.exists(), field.isNaN());
        case IS_NULL:
          return and(field.exists(), field.isNull());
        case IS_NOT_NAN:
          return and(field.exists(), field.isNotNaN());
        case IS_NOT_NULL:
          return and(field.exists(), field.isNotNull());
        default:
          // Handle OPERATOR_UNSPECIFIED and UNRECOGNIZED cases as needed
          throw new IllegalArgumentException("Unsupported operator: " + unaryFilter.getOperator());
      }
    } else {
      // Handle other FilterInternal types as needed
      throw new IllegalArgumentException("Unsupported filter type: " + f.getClass().getName());
    }
  }

  @InternalApi
  static Pipeline toPaginatedPipeline(
      Pipeline pipeline,
      Cursor start,
      Cursor end,
      Integer limit,
      LimitType limitType,
      Integer offset) {
    throw new UnsupportedOperationException(
        "Converting to pagination pipeline is not support yet.");
  }

  @InternalApi
  static AliasedAggregate toPipelineAggregatorTarget(AggregateField f) {
    String operator = f.getOperator();
    String fieldPath = f.getFieldPath();

    switch (operator) {
      case "sum":
        return Field.ofServerPath(fieldPath).sum().as(f.getAlias());

      case "count":
        return countAll().as(f.getAlias());
      case "average":
        return Field.ofServerPath(fieldPath).avg().as(f.getAlias());
      default:
        // Handle the 'else' case appropriately in your Java code
        throw new IllegalArgumentException("Unsupported operator: " + operator);
    }
  }

  @InternalApi
  static BooleanExpr toPipelineExistsExpr(AggregateField f) {
    String fieldPath = f.getFieldPath();

    if (fieldPath.isEmpty()) {
      return null;
    }
    return Field.ofServerPath(fieldPath).exists();
  }

  @InternalApi
  public static Map<String, Expr> selectablesToMap(Selectable... selectables) {
    Map<String, Expr> projMap = new HashMap<>();
    for (Selectable proj : selectables) {
      if (proj instanceof Field) {
        Field fieldProj = (Field) proj;
        projMap.put(fieldProj.getPath().getEncodedPath(), fieldProj);
      } else if (proj instanceof AliasedExpr) {
        AliasedExpr aliasedExpr = (AliasedExpr) proj;
        projMap.put(aliasedExpr.getAlias(), aliasedExpr.getExpr());
      }
    }
    return projMap;
  }

  @InternalApi
  public static Map<String, Expr> fieldNamesToMap(String... fields) {
    Map<String, Expr> projMap = new HashMap<>();
    for (String field : fields) {
      projMap.put(field, field(field));
    }
    return projMap;
  }
}

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/PipelineUtils.kt
@file:JvmName("PipelineUtils")
package com.google.cloud.firestore

import com.google.cloud.firestore.Query.ComparisonFilterInternal
import com.google.cloud.firestore.Query.CompositeFilterInternal
import com.google.cloud.firestore.Query.FilterInternal
import com.google.cloud.firestore.Query.LimitType
import com.google.cloud.firestore.Query.UnaryFilterInternal
import com.google.cloud.firestore.pipeline.AggregatorTarget
import com.google.cloud.firestore.pipeline.Constant
import com.google.cloud.firestore.pipeline.Field
import com.google.cloud.firestore.pipeline.Function
import com.google.cloud.firestore.pipeline.Function.Companion.countAll
import com.google.cloud.firestore.pipeline.Function.Companion.not
import com.google.firestore.v1.Cursor
import com.google.firestore.v1.StructuredQuery

internal fun toPipelineFilterCondition(f: FilterInternal): Function.FilterCondition {
  return when (f) {
    is ComparisonFilterInternal -> {
      when (f.operator) {
        StructuredQuery.FieldFilter.Operator.OPERATOR_UNSPECIFIED -> {
          TODO()
        }
        StructuredQuery.FieldFilter.Operator.LESS_THAN -> {
          Field.of(f.fieldReference.fieldPath).lessThan(f.value)
        }
        StructuredQuery.FieldFilter.Operator.LESS_THAN_OR_EQUAL -> {
          Field.of(f.fieldReference.fieldPath).lessThanOrEqual(f.value)
        }
        StructuredQuery.FieldFilter.Operator.GREATER_THAN -> {
          Field.of(f.fieldReference.fieldPath).greaterThan(f.value)
        }
        StructuredQuery.FieldFilter.Operator.GREATER_THAN_OR_EQUAL -> {
          Field.of(f.fieldReference.fieldPath).greaterThanOrEqual(f.value)
        }
        StructuredQuery.FieldFilter.Operator.EQUAL -> {
          Field.of(f.fieldReference.fieldPath).equal(f.value)
        }
        StructuredQuery.FieldFilter.Operator.NOT_EQUAL -> {
          not(Field.of(f.fieldReference.fieldPath).equal(f.value))
        }
        StructuredQuery.FieldFilter.Operator.ARRAY_CONTAINS -> {
          Field.of(f.fieldReference.fieldPath).arrayContains(f.value)
        }
        StructuredQuery.FieldFilter.Operator.IN -> {
          Function.In(
            Field.of(f.fieldReference.fieldPath),
            f.value?.arrayValue?.valuesList?.map { Constant.of(it) } ?: emptyList<Constant>(),
          )
        }
        StructuredQuery.FieldFilter.Operator.ARRAY_CONTAINS_ANY -> {
          Function.ArrayContainsAny(
            Field.of(f.fieldReference.fieldPath),
            f.value?.arrayValue?.valuesList?.map { Constant.of(it) } ?: emptyList<Constant>(),
          )
        }
        StructuredQuery.FieldFilter.Operator.NOT_IN -> {
          not(
            Function.In(
              Field.of(f.fieldReference.fieldPath),
              f.value?.arrayValue?.valuesList?.map { Constant.of(it) } ?: emptyList<Constant>(),
            )
          )
        }
        StructuredQuery.FieldFilter.Operator.UNRECOGNIZED -> {
          TODO()
        }
      }
    }
    is CompositeFilterInternal -> {
      when (f.operator) {
        StructuredQuery.CompositeFilter.Operator.OPERATOR_UNSPECIFIED -> {
          TODO()
        }
        StructuredQuery.CompositeFilter.Operator.AND -> {
          Function.And(f.filters.map { toPipelineFilterCondition(it) })
        }
        StructuredQuery.CompositeFilter.Operator.OR -> {
          Function.Or(f.filters.map { toPipelineFilterCondition(it) })
        }
        StructuredQuery.CompositeFilter.Operator.UNRECOGNIZED -> {
          TODO()
        }
      }
    }
    is UnaryFilterInternal -> {
      when (f.operator) {
        StructuredQuery.UnaryFilter.Operator.IS_NAN -> Field.of(f.fieldReference.fieldPath).isNaN()
        StructuredQuery.UnaryFilter.Operator.IS_NULL ->
          Field.of(f.fieldReference.fieldPath).isNull()
        StructuredQuery.UnaryFilter.Operator.IS_NOT_NAN ->
          not(Field.of(f.fieldReference.fieldPath).isNaN())
        StructuredQuery.UnaryFilter.Operator.IS_NOT_NULL ->
          not(Field.of(f.fieldReference.fieldPath).isNull())
        StructuredQuery.UnaryFilter.Operator.OPERATOR_UNSPECIFIED -> TODO()
        StructuredQuery.UnaryFilter.Operator.UNRECOGNIZED -> TODO()
      }
    }
    else -> {
      TODO()
    }
  }
}

internal fun toPaginatedPipeline(
  pipeline: Pipeline,
  start: Cursor?,
  end: Cursor?,
  limit: Int?,
  limitType: LimitType?,
  offset: Int?,
): Pipeline {
  var paginate: PaginatingPipeline = pipeline.paginate(limit ?: Int.MAX_VALUE)

  start?.let { paginate = setStartCursor(paginate, it) }
  end?.let { paginate = setEndCursor(paginate, it) }
  offset?.let { paginate = paginate.offset(it) }

  return limitType?.let {
    when (it) {
      LimitType.First -> paginate.firstPage()
      LimitType.Last -> paginate.lastPage()
    }
  } ?: paginate.firstPage()
}

internal fun toPipelineAggregatorTarget(f: AggregateField): AggregatorTarget {
  return when (f.operator) {
    "sum" -> {
      Field.of(f.getFieldPath()).sum().toField(f.alias)
    }
    "count" -> {
      countAll().toField(f.alias)
    }
    "avg" -> {
      Field.of(f.getFieldPath()).avg().toField(f.alias)
    }
    else -> TODO()
  }
}
